### ISSUE TITLE
include a Dockerfile

### ISSUE THREAD

**Issue by lavvy** [ffplayout/ffplayout-engine#20]
This is obviously a young project but looks ambitious. It's simple, light and looks straightforward. It uses ffmpeg which can handle video and audio unlike other radio stations out there and obviously run in the cloud. Its quite awesome.
I see how straight forward it is but it has a many of its piceses scattered all over the place likt the engine, gui, pipeline etc. I have spent about a day trying to stand it up in an alpine instance but no luck.
Even though i know it works but i just couldn't get it work.
I believe this only needs a lamp,  python3 and ffmpeg to work, but  i was not able to sit the parts in the right locations.
A simple dockerfile can put all the parts in the right locations (alpine preferable)

**Comment by jb-alvarado** [comment]
Yes a docker file would make it for some people more easy, but I'm not so experienced with docker. The Dockerfile would need:
- two linked folders from host
- a linked config file
- a init process to start ffplayout and the lamp stack

I have to figure out, how this works. But in the next months it looks a bit bad.

To your post: For your first test you only need python3, ffplayout-engine and ffmpeg (with libfdk_aac, or you need to change audio codec to aac). 
The GUI is a bit outdated, one time I would like to write it new.

**Comment by lavvy** [comment]
i dont even know maybe the dockerfile will better suit the ffplayout-gui project ffplayout/ffplayout-gui#1

**Comment by lavvy** [comment]
Ok i see you dont want to merge the two projects together. Thats fine. In that case we probably dont need lamp so the dockerfile will be pretty simpler.

FROM alpine:3.9
RUN apk add python3 ffmpeg nano
RUN mkdir /media/filler
ADD ffplayout.py /usr/local/bin
ADD . /home
CMD python3 ffplayout.py -l ~/ffplayout.log -f ~/playlist.json

not tested but pretty much what u need

**Comment by jb-alvarado** [comment]
Including both GUI and engine in a Docker container is a different story, this we can do. I only don't want do merge both project on github.

The problem is, the gui needs some improvements, and for that it makes more sense to write it totally new. At the moment I also have no good preview solution for the gui. I use flash video for that, but soon no browser will support flash anymore, so I need a different way. I thought webrtc could be nice, or HTTP-FLV when it latency is low.

**Comment by lavvy** [comment]
Its ok that you want to rewrite the gui. I see a lot of potentials in this approach cos of the technologies used. In my opinion, do you ever consider using python for the web gui like flask or tornado. I think localizing the gui will expose more features of the server to the web ui. For instance with websocket and webrtc you can do transitions, switching, selections of streams or videos pretty much like desktop apps. And with python engine and python gui there would be consistencies and reuse of classes between the engine and the gui.

**Comment by lavvy** [comment]
i have still not been able to set this thing up, i noticed that stock ffmpeg wont work bcos it doesnt come with libfdk-aac therefore ffmpeg has to be compiled from source. https://github.com/alfg/docker-ffmpeg/blob/master/Dockerfile#L57. 
And its complaining of "Too many packets buffered for output stream 0:1" which i think has to do with the pipebuffer project(that i dont know how to combine)
i am not able to get this working i should be able to create a dockerfile for it but its really kicking my butts .

**Comment by jb-alvarado** [comment]
No forget pipebuffer :)! I have use it in the past, but now it is not necessary anymore. You also don't need libfdk-aac, you can replace this with `aac`, a almost new ffmpeg build (4.*) is totally enough. 

I would recommend to test it first on a desktop system, with ffmpeg, ffprobe and ffplay installed:
- Set `preview` to **True** 
- prepare a playlist, or use watchfolder
- run it with: `./ffplayout  -l ~/ffplayout.log`

When this works you get an idea how it should run, and then you can transfer it to your server.

To the  packets buffering I can not say much, at the moment, I never had this issue. For this I need more infos.

And to your post before:
Yes I though about to use django for the GUI backend and vue.js/bootstrap for the frontend. 
For me the GUI needs:
- a configuration page, where you can configure the engine and some gui staff
- a upload function for adding new clips to the storage
- login and user management
- modulation, to make it more simple to add custom plugins

But for all this I have no time at the moment.

**Comment by lavvy** [comment]
So where is the default watchfolder? I can't see it in the conf file. Secondly I tried to use the sample playlist json on the readme but it didn't work. I think the format is the issue

**Comment by lavvy** [comment]
please disregard my last comment . i have gotten the watchfolder config settings. Its working now . i will test it well and try and submit a pull request for  docker

**Comment by jb-alvarado** [comment]
The folder you can set here: https://github.com/ffplayout/ffplayout-engine/blob/3eac6dbfb6fad604cb852ea1cc0a5cc875bc46c7/ffplayout.conf#L102

Can you describe more, what did not work? What says the log file, and the console output?

**Comment by lavvy** [comment]
i have gotten the watchfolder to work but i have not gotten the playlist.json to work. i used the sample on the readme page but it doesnt seem to be be recognized and no logs printed.
python3 ffplayout.py -l ffplayout.log -f playlist.json

**Comment by lavvy** [comment]
i have now been able to get the example playlist.json in the readme to work. It was a format error . I will send a pull request to correct it

**Comment by lavvy** [comment]
I think for now docker could be an overkill especially without gui. I am thinking a simple install script or pip3 install could be ideal. 
It was a bit confusing before but now its now more streamline particularly by changing the contents related directories to /media. 
Running it now as a docker (without gui) will pose a little problem bcos it requires some user's specific config before starting.  
So i think when the gui is ready we can package a dockerfile for it and then the user can then add his configs and contents on the fly while the server runs .

**Comment by jb-alvarado** [comment]
Yes I think this to. The engine can run with low privilege, have no much dependencies and open no security holes. So it really makes no sense to put it in a container. With GUI is another story, there it really makes sense.

**Comment by jb-alvarado** [comment]
Just for people how end up here: https://github.com/ffplayout/ffplayout-docker
